README.md
# capture.js

A library for capturing your DOM.

[![npm version](https://badge.fury.io/js/mobify-capturejs.svg)](http://badge.fury.io/js/mobify-capturejs)

## How to use

Right now, there are no instructions on how to use capture.js standalone. Please
go to www.mobifyjs.com for information on how to use capturing.

## How to create and ship a change

### Install the dependencies:

Get NodeJS v8.10 with npm v5.7: https://nodejs.org/en/download/

```bash
npm install
$(npm bin)/bower install
```

### Run the build with a watcher

Then, run the following command to watch and build the capture.js file as you
develop:

```bash
$(npm bin)/grunt serve
```

### Running tests

To run the tests in your browser

```
$(npm bin)/grunt serve
open http://localhost:3000/tests/index.html
```

To run the tests from the commandline

```
$(npm bin)/grunt test
```

To run the tests across several browsers using suacelabs
```
export SAUCE_USERNAME=<saucelabs username>
export SAUCE_ACCESS_KEY=<saucelabs access key>
$(npm bin)/grunt saucelabs
```
You can monitor test progress via the suacelabs dashbaord:
[https://app.saucelabs.com/dashboard/tests](https://app.saucelabs.com/dashboard/tests)

The test suite will likely take several minutes to run across the test matrix
defined in `Gruntfile.js`

Then, do the following to ship the change:

* Make your code changes
* Ensure the tests still work in your browser, on the commandline, in
  continuous integration and in Sauce Labs
* Create a pull-request
* Get your change reviewed and :+1:'ed

## Where to get help

Talk to @stellafang, @noahadams or @johnboxall


CHANGELOG
1.0.7
    - Upgrade dependencies (grunt-conbtrib-qunit and grunt-saucelabs).
    - Make bower, browserify, and grunt-cli dependencies explicit.
    - Update list of browsers to run saucelabs on.
    - Remove plaintext from sourceDoc after captured doc elements have been
      saved into the capturedDoc to prevent googlebot from picking up the
      lingering plaintext as metadata.
    - Add checks to tests to confirm that plaintext was added before
      Capture.init and removed after the callback in Capture.init.
1.0.6
    - Fixed issue where <!-- --> comments containing '<head>' created false
      positives when looking for the <head> section in HTML
1.0.5
    - iOS8+ scroll fix is applied only when rendering an adapted document,
      resolving an issue with meta viewport tags and restored documents (PR #11)
1.0.4
    - Fix scroll issue regression with iOS9 (PR #9)
    - Rename package to mobify-capturejs to avoid conflict with capturejs npm
      package

1.0.3
    - Fix issues with having <body> in a script tag above the real
      body tag.
1.0.2
    - iOS8 smart banner fix
1.0.1
    - Changes necessary for adaptive.js
    - Remove adaptive.js and loader.js from capture
    - Set Capture to the window
1.0.0
    - Initial commit moving capture.js out of the mobifyjs repo


tests/patchAnchorLinks-tests.js
require(['saucelabs-config', 'qunit', 'patchAnchorLinks'], function(_, QUnit, patchAnchorLinks) {
    module("Anchor");
    var skip = function() {
        // Test harness does not work on Android 4.0.x and less because
        // of an iframe scrolling bug. However, the functionality works
        // but needs to be manually tested.
        // Also we don't run these tests in iOS8 because you can't seem
        // to query for the scroll position in iframes.
        return /(android.*(4\.0\.|(2|3)\.\d)|ip(hone|od|ad).*Version\/8.0)|/i.test(window.navigator.userAgent);
    };

    var testSetup = function(id, ready) {
        var opts = {
            id: "anchor-test",
            src: "/tests/fixtures/anchor-test.html"
        };
        var $iframe = $("<iframe>", opts);
        var el = $iframe[0];

        window.addEventListener("message", function onMessage(event) {
            if (event.source != el.contentWindow) return;
            window.removeEventListener("message", onMessage, false);
            var win = el.contentWindow;
            var doc = el.contentWindow.document;

            var triggerClick = function (el) {
                // We have to use this instead of `el.click`, because
                // `el.click` seems to be broken in PhantomJS
                // http://stackoverflow.com/questions/17468611/triggering-click-event-phantomjs

                var e = doc.createEvent("MouseEvent");
                e.initMouseEvent(
                    "click",
                    true,
                    true,
                    win,
                    null,
                    0,0,0,0,
                    false, false, false, false,
                    0,
                    null
                );
                el.dispatchEvent(e);
            }
            // Set an offset so we can test scrolling
            win.scrollTo(0, 10);

            // Click the link in question
            var link = doc.getElementById(id);
            triggerClick(link);

            setTimeout(function() {
                if (skip()) {
                    ok(true, "Test harness does not work on Android < 4.0. Must be tested manually.");
                    return start();
                }

                // Call ready so we can fire assertions
                ready(win, doc);
            }, 100);
        }, false);

        $("#qunit-fixture").append($iframe);
    };

    var getScrollTop = function(win) {
        return typeof win.scrollY !== "undefined" ? win.scrollY : win.document.documentElement.scrollTop;
    };

    asyncTest("Regular Anchor Tag to Element", function() {
        testSetup('test1', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            
            // Some browsers have a little different idea about how
            // much do you need to do to scroll them in to view.
            var scrollTop = getScrollTop(win);
            ok((scrollTop >= 490) && (scrollTop <= 510), "Expected scroll position to be between [490, 510], got " + scrollTop);

            start();
        });
    });

    asyncTest("Regular Anchor Tag to Top of Page", function() {
        testSetup('test2', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            equal(getScrollTop(win), 0);

            start();
        });
    });

    asyncTest("Anchor Tag with Handler (no Prevent Default)", function() {
        testSetup('test3', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            
            // Some browsers have a little different idea about how
            // much do you need to do to scroll them in to view.
            var scrollTop = getScrollTop(win);
            ok((scrollTop >= 490) && (scrollTop <= 510), "Expected scroll position to be between [490, 510], got " + scrollTop);
            
            ok(win.testFired3, "Click handler was fired.");

            start();
        });
    });

    asyncTest("Anchor Tag with Handler (Prevent Default)", function() {
        testSetup('test4', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            equal(getScrollTop(win), 10);
            ok(win.testFired4, "Click handler was fired.");

            start();
        });
    });

    asyncTest("Anchor Tag with onclick (no Prevent Default)", function() {
        testSetup('test5', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            
            // Some browsers have a little different idea about how
            // much do you need to do to scroll them in to view.
            var scrollTop = getScrollTop(win);
            ok((scrollTop >= 490) && (scrollTop <= 510), "Expected scroll position to be between [490, 510], got " + scrollTop);

            ok(win.testFired5, "Click handler was fired.");

            start();
        });
    });

    asyncTest("Anchor Tag with Handler (return void)", function() {
        testSetup('test6', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            equal(getScrollTop(win), 10);
            ok(win.testFired6, "Click handler was fired.");

            start();
        });
    });

    asyncTest("Anchor Tag with Handler (calls Prevent Default)", function() {
        testSetup('test7', function(win, doc) {
            equal(win.hadHashOnLoad, false);
            equal(getScrollTop(win), 10);
            ok(win.testFired7, "Click handler was fired.");

            start();
        });
    });
    test("Old Firefox UA detection", function() {
        firefoxUAs = {
            "3.6.9": "Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.2.9) Gecko/20100915 Gentoo Firefox/3.6.9",
            "25.0": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:25.0) Gecko/20100101 Firefox/25.0",
            "29.0": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:29.0) Gecko/20100101 Firefox/29.0"
        };
        ok(patchAnchorLinks._isOldFirefox(firefoxUAs["3.6.9"]), "firefox 3.6.9 is old");
        ok(patchAnchorLinks._isOldFirefox(firefoxUAs["25.0"]), "firefox 25 is also old");
        ok(!patchAnchorLinks._isOldFirefox(firefoxUAs["29.0"]), "firefox 29 is not old!");
    });
});


web.js
/**
 * Server for testing the number of requests made by browsers using capturing.
 *
 * Provides "counter" views for resources, including the ability to simulate slow
 * and cached responses.
 */
var express = require('express');

var mimeMap = {
    'js': 'application/javascript',
    'css': 'text/css',
    'jpg': 'image/jpg',
}

var counters = {};

// Views

var report = function(req, res) {
    var prefix = req.params.prefix;
    res.jsonp(counters[prefix]);
}

var increment = function(req, res) {
    var prefix = req.params.prefix;
    var key = req.params.key;

    var counter = counters[prefix] = counters[prefix] || {};
    if (!counter[key]) {
        counter[key] = 1;
    } else {
        counter[key]++;
    }

    // Correct mime to avoid console warnings.
    var extension = req.path.split('.').pop(-1) || 'js';
    res.append('Content-Type', mimeMap[extension]);

    res.send('/* ' + counter[key] + ' */');
};

var incrementSlow = function(req, res) {
    setTimeout(function() {
        increment(req, res);
    }, 1000);
};

var incrementCached = function(req, res) {
    res.append('Cache-Control', 'max-age=3600');
    increment(req, res);
};

var index = function(req, res) {
    res.render('index.html', {prefix: +new Date});
};

// App Setup

var app = express();

// Remove headers that may interfere with caching.
app.disable('etag');
app.disable('x-powered-by');

app.engine('.html', require('ejs').__express);
app.set('view engine', 'html');

app.use(express.static(__dirname + '/public'));

app.get('/slow/:prefix/:key', incrementSlow);
app.get('/cache/:prefix/:key', incrementCached);
app.get('/:prefix/:key', increment);
app.get('/:prefix/', report);
app.get('/', index);

var srv = app.listen(process.env.PORT || 3000, '0.0.0.0', function() {
    var addr = srv.address()
    console.log('http://%s:%d/', addr.address, addr.port);
});

public/report.js
window.report = function(counts) {
    var preparserMakesRequests = false;
    var duplicatesRequestsForCachedResponses = false;
    var terminatesRequests = false;

    for (var key in counts) {
        if (counts.hasOwnProperty(key)) {
            if (counts[key] > 1) {
                preparserMakesRequests = true;
            }

            if (/cache/.test(key) && counts[key] > 1) {
                duplicatesRequestsForCachedResponses = true;
            }
        }
    }

    var text = preparserMakesRequests ? "YES" : "NO";
    var color = preparserMakesRequests ? "red" : "green";
    var el = document.createElement("code");
    el.innerText = text;
    el.style.color = color;
    console.log(el.style);
    document.getElementById("preparser-requests").appendChild(el);

    var text = duplicatesRequestsForCachedResponses ? "YES" : "NO";
    var color = duplicatesRequestsForCachedResponses ? "red" : "green";
    var el = document.createElement("code");
    el.innerText = text;
    el.style.color = color;
    document.getElementById("duplicates-requests").appendChild(el);

    var el = document.createElement('pre');
    el.innerText = 'Request Counts for ' + prefix + ':\n' + JSON.stringify(counts, null, 2);
    document.body.appendChild(el);
}

document.addEventListener('DOMContentLoaded', function() {
    var script = document.createElement('script');
    script.src = '/' + window.prefix + '/?callback=report';
    document.body.appendChild(script);
}, false);


src/patchAnchorLinks.js
// Fixes anchor links (on FF)
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(["mobifyjs/utils"], factory);
    } else if (typeof exports === 'object') {
        // Node. Does not work with strict CommonJS, but
        // only CommonJS-like environments that support module.exports,
        // like Node.
        var Utils = require('../../mobifyjs-utils/utils.js');
        module.exports = factory(Utils);
    } else {
        // Browser globals (root is window)
        root.Utils = factory(root.Utils);
    }
}(this, function (Utils) {
    var exports = {};

    var isOldFirefox = function(ua) {
        ua = ua || window.navigator.userAgent;
        var match = /(firefox|fennec)[\/\s](\d+)/i.exec(ua);
        if (match) {
            var version = +match[2];
            if (version >= 29) {
                return false;
            }
            return true;
        }

        return false;
    };

    var _patchAnchorLinks = function(doc) {
        // Anchor links in FF, after we do `document.open` cause a page
        // navigation (a refresh) instead of just scrolling the
        // element in to view.
        //
        // So, we prevent the default action on the element, and
        // then manually scroll it in to view (unless some else already
        // called prevent default).

        var body = doc.body;

        if (!(body && body.addEventListener)) {
            // Body is not there or we can't bind as expected.
            return;
        }

        var _handler = function(e) {
            // Handler for all clicks on the page, but only triggers
            // on proper anchor links.

            var target = e.target;

            var matches = function(el) {
                return (el.nodeName == "A") && (/^#/.test(el.getAttribute('href')));
            };

            if (!matches(target)) {
                return;
            }
            
            // Newer browsers support `e.defaultPrevented`. FF 4.0 supports `e.getPreventDefault()`
            var defaultPrevented = (typeof e.defaultPrevented !== "undefined") ?
                e.defaultPrevented :
                e.getPreventDefault && e.getPreventDefault();

            if (!defaultPrevented) {
                // Prevent the default action, which would cause a
                // page refresh.
                e.preventDefault();

                // But pretend that we didn't call it.
                e.defaultPrevented = false;

                // We have to wait and see if anyone else calls
                // `preventDefault`. If they do, we don't scroll.
                var scroll = true;

                // Override the `preventDefault` to stop  us from scrolling.
                e.preventDefault = function() {
                    e.defaultPrevented = true;
                    scroll = false;
                };

                // If no other events call `preventDefault` we manually
                // scroll to the element in question.
                setTimeout(function() {
                    if (scroll) {
                        _scrollToAnchor(target.getAttribute('href'));
                    }
                }, 50);
            }
        };


        var _scrollToAnchor = function(anchor) {
            // Scrolls to the element, if any, that matches
            // the given anchor link (eg, "#foo").

            var anchorRe = /^#([^\s]*)/;
            var match = anchor.match(anchorRe);
            var target;
            
            // Find the target, if any
            if (match && match[1] === "") {
                target = doc.body;
            } else if (match && match[1]) {
                target = doc.getElementById(match[1]);
            }

            // Scroll to it, if it exists
            if (target) {
                target.scrollIntoView && target.scrollIntoView();
            }
        };

        // We have to get the event through bubbling, otherwise
        // events cancelled by the return value of an onclick
        // handler are not correctly handled.
        body.addEventListener('click', _handler, false);
    };

    var patchAnchorLinks = function() {
        if (!isOldFirefox()) {
            return;
        }

        Utils.waitForReady(document, _patchAnchorLinks);
    };

    patchAnchorLinks._isOldFirefox = isOldFirefox;


    return patchAnchorLinks;
}));


Gruntfile.js
module.exports = function(grunt) {

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        express: {
            capturejs: {
                options: {
                    hostname: '0.0.0.0',
                    port: 3000,
                    base: '.',
                    debug: true
                }
            },
            test: {
                options: {
                    hostname: '0.0.0.0',
                    port: 3001,
                    base: '.',
                    debug: true
                }
            }
        },
        qunit: {
            capturejs: {
              options: {
                timeout: 20000,
                urls: [
                  'http://localhost:3001/tests/index.html',
                ]
              }
            }
        },
        browserify: {
            capturejs: {
                src: ['src/capture.js'],
                dest: 'build/capture.js'
            },
        },
        uglify: {
            capturejs: {
                files: {
                    'build/capture.min.js': ['build/capture.js']
                }
            },
        },
        watch: {
            files: ["src/**/*.js"
            ],
            tasks: ['build'],
        },
        'saucelabs-qunit': {
            capturejs: {
                options: {
                    // Set the Saucelabs username and key in your environment variables
                    // by setting SAUCE_USERNAME and SAUCE_ACCESS_KEY
                    urls: [
                        'http://localhost:3001/tests/index.html'
                    ],
                    testname: 'CaptureJS QUnit Tests',
                    concurrency: 16,
                    tunneled: true,
                    detailedError: true,
                    // Poll for status every 10 seconds for 10 minutes
                    pollInterval: 10000,
                    statusCheckAttempts: 60,
                    // For available browsers and how to configure them see
                    // https://saucelabs.com/docs/platforms and
                    // https://wiki.saucelabs.com/display/DOCS/Platform+Configurator#/
                    browsers: [
                        { // IE 10 on Windows 8
                            browserName: 'internet explorer',
                            platform: 'Windows 2012',
                            version: '10'
                        },
                        { // IE 11 on Windows 8.1
                            browserName: 'internet explorer',
                            platform: 'Windows 8.1',
                            version: '11'
                        },
                        { // Latest IE Edge on Windows 10
                            browserName: 'MicrosoftEdge',
                            platform: 'Windows 10'
                        },
                        { // Latest Chrome on Windows 10
                            browserName: 'chrome',
                            platform: 'Windows 10'
                        },
                        { // Latest Chrome on Linux (unknown distro)
                            browserName: 'chrome',
                            platform: 'Linux'
                        },
                        { // Latest Chrome on macOS 10.13
                            browserName: 'chrome',
                            platform: 'macOS 10.13'
                        },
                        { // Lowest known working version of FF
                            platform: 'Windows 10',
                            browserName: 'firefox',
                            version: '4.0'
                        },
                        { // Highest known working version of FF on Windows
                            platform: 'Windows 10',
                            browserName: 'firefox',
                        },
                        {   // Newest iOS simulator as of 2019-04-02
                            browserName: 'ipad',
                            platform: 'macOS 10.13',
                            version: '12.0'
                        },
                        {   // Oldest iOS simulator as of 2019-04-02
                            browserName: 'ipad',
                            platform: 'macOS 10.11',
                            version: '9.3'
                        },
                        {   // Newest iOS simulator as of 2019-04-02
                            browserName: 'iphone',
                            platform: 'macOS 10.13',
                            version: '12.0'
                        },
                        {   // Oldest iOS simulator as of 2019-04-02
                            browserName: 'iphone',
                            platform: 'macOS 10.11',
                            version: '9.3'
                        },
                        { // Android 4.4 Android Browser (simulator)
                            browserName: 'android',
                            platform: 'Linux',
                            version: '4.4'
                        },
                        { // Android 8.0 Chrome? (simulator)
                            browserName: 'android',
                            platform: 'Linux',
                            version: '8.0'
                        },
                    ]
                }
            }
        },
        aws_s3: {
            capturejs: {
                options: {
                    bucket: 'mobify',
                    params: {
                        CacheControl: "public,max-age=1800", // cache for 30 min
                    }
                },
                files: [
                    { // unminified dev build
                        src: ["build/capture.js"],
                        dest: "capturejs/capture-<%= pkg.version %>.js",
                    },
                    { // unminified dev build to latest
                        src: ["build/capture.js"],
                        dest: "capturejs/capture-latest.js",
                    },
                    { // minified production build
                        src: ["build/capture.min.js"],
                        dest: "capturejs/capture-<%= pkg.version %>.min.js",
                    },
                    { // minified production build to latest
                        src: ["build/capture.min.js"],
                        dest: "capturejs/capture-latest.min.js",
                    }
                ]
            }
        }
    });

    grunt.loadNpmTasks('grunt-express');
    grunt.loadNpmTasks('grunt-contrib-qunit');
    grunt.loadNpmTasks('grunt-aws-s3');
    grunt.loadNpmTasks('grunt-saucelabs');
    grunt.loadNpmTasks('grunt-browserify');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.registerTask('build', ['browserify', 'uglify']);
    grunt.registerTask('saucelabs', ['test', 'saucelabs-qunit']);
    grunt.registerTask('test', ['build', 'express:test', 'qunit']);
    grunt.registerTask('serve', ['build', 'express:capturejs', 'watch']);
    grunt.registerTask('deploy', ['build', 'aws_s3']);
};
